testsuite/cluster-test/src/atomic_histogram.rs

```diff
@@ -0,0 +1,182 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+#![forbid(unsafe_code)]
+
+use crate::tx_emitter::TxStats;
+use std::sync::atomic::{AtomicU64, Ordering};
+
+const HISTOGRAM_CAPACITY: usize = 2048;
+const HISTOGRAM_STEP_MILLIS: u64 = 50;
+
+pub struct AtomicHistogram {
+    capacity: usize,
+    step_millis: u64,
+    buckets: Vec<AtomicU64>,
+}
+
+impl Default for AtomicHistogram {
+    fn default() -> AtomicHistogram {
+        let mut buf = Vec::with_capacity(HISTOGRAM_CAPACITY);
+        for _i in 0..HISTOGRAM_CAPACITY {
+            buf.push(AtomicU64::new(0));
+        }
+        Self {
+            capacity: HISTOGRAM_CAPACITY,
+            step_millis: HISTOGRAM_STEP_MILLIS,
+            buckets: buf,
+        }
+    }
+}
+
+impl AtomicHistogram {
+    pub fn to_vec(&self) -> Vec<u64> {
+        let mut buf = Vec::with_capacity(self.capacity);
+        for i in 0..self.capacity {
+            buf.push(self.buckets[i].load(Ordering::Relaxed));
+        }
+
+        buf
+    }
+
+    fn get_bucket_num(&self, latency: u64) -> u64 {
+        let bucket_num = latency / self.step_millis;
+        if bucket_num >= self.capacity as u64 - 2 {
+            return self.capacity as u64 - 1;
+        }
+        bucket_num
+    }
+
+    pub fn record_latency_ms(&self, latency_ms: u64, num_requests: u64) {
+        let bucket_num = self.get_bucket_num(latency_ms);
+        self.buckets[bucket_num as usize].fetch_add(num_requests as u64, Ordering::Relaxed);
+    }
+
+    pub fn sub(a: &[u64], b: &[u64]) -> Vec<u64> {
+        let mut buf = Vec::with_capacity(a.len());
+        if a.len() != b.len() {
+            for &i in a.iter() {
+                buf.push(i);
+            }
+        } else {
+            for i in 0..a.len() {
+                buf.push(a[i] - b[i]);
+            }
+        }
+        buf
+    }
+
+    pub fn cal_percentile_latency(numerator: u64, denominator: u64, stats: &TxStats) -> u64 {
```

Comment:
I think this api got little bit mixed up.
We should not reference tx_emitter classes from here. Imagine histogram to be as a utility class that anyone can use. Maybe it will get moved outside of cluster test crate even.

I think what you want to do probably is to split this into two classes:
`AtomicHistogramAccumulator` that will hold vec<AtomicU64> and `AtomicHistogramSnapshot` that will hold vec<u64>.
Then existing API will change as
- instead of to_vec, AtomicHistogramAccumulator will have .snapshot() method.
- cal_percentile_latency will be on AtomicHistogramSnapshot
- Instead of `sub(a: &[u64], b: &[u64]) ` would would implement trait `Sub` on AtomicHistogramSnapshot
- In tx_emitter, `StatsAccumulator` will hold AtomicHistogramAccumulator and `TxStats` will hold `AtomicHistogramSnapshot`